arcgisUM.py

```diff
@@ -0,0 +1,206 @@
+    logger.debug("group search string: original: {}".format(searchString))
+
+    # quote characters that are special in the group search.
+    searchString = searchString.translate(str.maketrans({"?":  r"\?","*":  r"\*"}))
+
+    logger.debug("group search string: escaped: {}".format(searchString))
+    
+    try:
+        gis_groups = arcGISAdmin.groups.search(searchString)
+    except RuntimeError as exp:
+        logger.error("arcGIS error finding group: {} exception: {}".format(searchString,exp))
+    
+    if len(gis_groups) > 0:
+        return gis_groups.pop()
+
+    return None
+
+
+def addCanvasUsersToGroup(instructorLog, group, courseUsers):
+    """Add new users to the ArcGIS group.  """
+    groupNameAndID = util.formatNameAndID(group)
+    
+    logger.info("addCanvasUsersToGroup: enter")
+    
+    if len(courseUsers) == 0:
+        logger.info('No new users to add to ArcGIS Group {}'.format(groupNameAndID))
+        return instructorLog
+
+    logger.info('Adding Canvas Users to ArcGIS Group {}: {}'.format(groupNameAndID, courseUsers))
+    # ArcGIS usernames are U-M uniqnames with the ArcGIS organization name appended.
+    user="NULL"
+    arcGISFormatUsers = formatUsersNamesForArcGIS(user, courseUsers)
+    logger.debug("addCanvasUsersToGroup: formatted: {}".format(arcGISFormatUsers))
+    
+    #results = group.addUsersToGroups(users=','.join(arcGISFormatUsers))
+    results = group.add_users(arcGISFormatUsers)
+    logger.debug("adding: results: {}".format(results))
+
+    usersNotAdded = results.get('notAdded')
+    """:type usersNotAdded: list"""
+    usersCount = len(arcGISFormatUsers)
+    usersCount -= len(usersNotAdded) if usersNotAdded else 0
+    logger.debug("usersCount: {}".format(usersCount))
+    logger.debug("aCUTG: instructorLog 1: [{}]".format(instructorLog))
+    instructorLog += 'Number of users added to group: [{}]\n\n'.format(usersCount)
+    logger.debug("aCUTG: instructorLog 2: [{}]".format(instructorLog))
+    if usersNotAdded:
+        logger.warning('Warning: Some or all users not added to ArcGIS group {}: {}'.format(groupNameAndID, usersNotAdded))
+        instructorLog += 'Users not in group (these users need ArcGIS accounts created for them):\n' + '\n'.join(['* ' + userNotAdded for userNotAdded in usersNotAdded]) + '\n\n' + 'ArcGIS group ID number:\n{}\n\n'.format(group.id)
+    instructorLog += '- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n'
+    logger.debug("aCUTG: instructorLog 3: [{}]".format(instructorLog))
+
+    logger.info("addCanvasUsersToGroup: instructorLog: [{}]".format(instructorLog))
+    return instructorLog
+
+
+def getCurrentArcGISMembers(group, groupNameAndID):
+    groupAllMembers = {}
+
+    try:
+        groupAllMembers = group.get_members()
+    except RuntimeError as exception:
+        logger.error('Exception while getting users for ArcGIS group "{}": {}'.format(groupNameAndID, exception))
+            
+    groupUsers = groupAllMembers.get('users')
```

Comment:
if an exception is thrown then will end up on line 137, then continues to line 139  then `groupAllMembers` is empty here, isn't the `groupAllMembers.get('users')` will again cause an exception if there is not return in the exception block?

Reply:
No.  GroupAllMembers has an empty dictionary.  The following get will return a None.

Reply:
Good question.